### PR TITLE
[Triton/Gluon] [Kernel] Add fused SiLU-and-multiply backward

### LINKED ISSUE #5541: Add fused SiLU-and-multiply backward kernel
## Problem

AITER provides fused forward implementations for the standard
`silu(gate) * up` operation, including `aiter.silu_and_mul`, but it does not
provide the corresponding fused backward operation. Training callers therefore
fall back to several framework elementwise kernels or maintain a downstream
implementation.

## Proposed change

Add a Triton `silu_and_mul_backward` operation that:

- computes gradients for concatenated `[gate, up]` inputs in one launch;
- evaluates the sigmoid and gradient products in FP32 before converting back;
- supports FP16, BF16, and FP32 inputs;
- supports strided incoming gradients and an optional preallocated output; and
- uses bounded two-dimensional tiles for wide training dimensions.

## Scope

This issue covers one backward kernel, its public wrapper, gfx942/gfx950
configuration, correctness tests, and benchmark. It does not add another
forward kernel; callers should reuse the existing `aiter.silu_and_mul` forward
operation.

### PR DESCRIPTION
**Pull request by DaiXindi-AMD**
## Summary

Add a fused Triton backward operation for `silu(gate) * up` with concatenated
`[gate, up]` input layout.

Closes #5541.

## Motivation

AITER already provides the SiLU-and-multiply forward operation, but training
currently builds its backward from multiple eager PyTorch operations. This PR
adds the matching fused backward while keeping the implementation in the
existing activation family.

## Changes

- Add `_silu_and_mul_backward_kernel` beside the existing SiLU-and-multiply
  forward kernel and reuse the shared `_sigmoid_exp2` device helper.
- Add `silu_and_mul_backward(grad_output, x, out=None)` to
  `aiter.ops.triton.activation`.
- Evaluate SiLU and both gradient products in FP32 before converting to the
  input dtype.
- Support FP16, BF16, and FP32, arbitrary leading dimensions, strided
  `grad_output`, optional preallocated output, empty rows, and non-current-GPU
  launches.
- Add a config-aware kernel `repr`, unit tests, a benchmark, and the measured
  gfx950 tuning configuration.
- Reuse a shared `select_leq_config` helper for upper-bound config tables and
  migrate the two existing local copies in `fused_clamp_act_mul` and
  `fused_add_rmsnorm_pad`.
- Route the touched config reads through `resolve_config_dir()` and document
  the shared selector in the config rulebooks.
- Correct the shared VGPR table parser for benchmark labels containing units.

## Hardware support and tuning

This PR supports **gfx950 only**. It does not ship or claim support for gfx942;
an unsupported architecture fails required-config resolution with the missing
architecture-specific path.

The measured config is stored at:

```text
aiter/ops/triton/configs/gfx950/triton/fusions/silu_and_mul_backward/DEFAULT.json
```

`width` below is the gate/up width, so `x.shape[-1] == 2 * width`.
`select_leq_config` chooses the smallest matching `N_LEQ_*` entry and otherwise
uses `any`:

| Width | `BLOCK_M` | `BLOCK_N` | `num_warps` | `num_stages` |
|---:|---:|---:|---:|---:|
| `<= 64` | 32 | 64 | 4 | 1 |
| `<= 128` | 32 | 128 | 4 | 1 |
| `<= 256` | 16 | 256 | 4 | 1 |
| `<= 512` | 8 | 512 | 4 | 1 |
| `> 512` | 4 | 1024 | 4 | 1 |

`BLOCK_M` and `BLOCK_N` are the row and column tiles. At launch, `BLOCK_M` is
capped by the next power of two of the row count; widths above 1024 use
multiple column tiles. `num_warps` and `num_stages` control launch scheduling.
The kernel `repr` records the effective `BLOCK_M` and `BLOCK_N` specialization.

## Performance

Measured at commit `74dcacab94842bac67e59b50e202ecea2b129672`.

- GPU: AMD Instinct MI350X VF (`gfx950`)
- ROCm: 7.2.4
- HIP runtime: 7.2.53211
- Python: 3.10.12
- PyTorch: `2.13.0+rocm7.2`
- Triton: 3.7.0
- dtype: BF16
- Timing: `triton.testing.do_bench`, `warmup=25 ms`, `rep=100 ms`, arithmetic mean

Commands:

```bash
for width in 4096 12288; do
  for metric in time bandwidth; do
    ROCR_VISIBLE_DEVICES=0 PYTHONPATH=. \
      python op_tests/op_benchmarks/triton/bench_silu_and_mul_backward.py \
      --width "${width}" --dtype bf16 --provider all --metric "${metric}" \
      --warmup 25 --rep 100
  done
done
```

Each provider is measured by an independent `do_bench` invocation. `aiter`
measures the allocation-inclusive public API (`out=None`),
`aiter-preallocated` allocates `out` before timing, and `torch` is an
FP32-intermediate eager implementation of the same derivative. There was no
previous AITER backward operator, so the eager implementation is the functional
baseline rather than a before/after AITER kernel baseline.

### Latency and speedup

| Rows | Width | Preallocated `out` (us) | `out=None` (us) | FP32 eager (us) | Speedup: preallocated / `out=None` |
|---:|---:|---:|---:|---:|---:|
| 1 | 4,096 | 10.095 | 7.806 | 89.599 | 8.88x / 11.48x |
| 128 | 4,096 | 6.817 | 6.939 | 87.115 | 12.78x / 12.55x |
| 2,048 | 4,096 | 32.894 | 33.041 | 234.537 | 7.13x / 7.10x |
| 8,192 | 4,096 | 107.258 | 107.228 | 1,032.173 | 9.62x / 9.63x |
| 1 | 12,288 | 11.821 | 10.054 | 104.233 | 8.82x / 10.37x |
| 128 | 12,288 | 9.110 | 11.659 | 104.813 | 11.50x / 8.99x |
| 2,048 | 12,288 | 89.151 | 89.169 | 757.152 | 8.49x / 8.49x |
| 8,192 | 12,288 | 256.765 | 256.638 | 3,079.848 | 11.99x / 12.00x |

Geometric-mean speedup over the FP32 eager reference is **9.73x** with a
preallocated output and **9.92x** through the allocation-inclusive public API.
For rows >= 2048, the two AITER paths differ by at most 0.45%; sub-percent
differences are measurement noise, not evidence that allocation improves the
kernel. Small cases are launch/API-latency bound.

### Effective bandwidth

Logical BF16 traffic is three input reads plus two output writes per gate/up
pair:

```text
5 * rows * width * sizeof(bfloat16) = 10 * rows * width bytes
```

Utilization uses AMD's published MI350X peak theoretical HBM bandwidth of
[8.0 TB/s per GPU](https://www.amd.com/en/products/accelerators/instinct/mi350/mi350x.html),
or 8,000 GB/s in decimal units.

| Rows | Width | Logical bytes | Preallocated GB/s (% peak) | `out=None` GB/s (% peak) | FP32 eager logical GB/s (% peak) |
|---:|---:|---:|---:|---:|---:|
| 1 | 4,096 | 40,960 | 4.1 (0.05%) | 5.2 (0.07%) | 0.5 (0.01%) |
| 128 | 4,096 | 5,242,880 | 769.1 (9.61%) | 755.5 (9.44%) | 60.2 (0.75%) |
| 2,048 | 4,096 | 83,886,080 | 2,550.2 (31.88%) | 2,538.9 (31.74%) | 357.7 (4.47%) |
| 8,192 | 4,096 | 335,544,320 | 3,128.4 (39.10%) | 3,129.3 (39.12%) | 325.1 (4.06%) |
| 1 | 12,288 | 122,880 | 10.4 (0.13%) | 12.2 (0.15%) | 1.2 (0.01%) |
| 128 | 12,288 | 15,728,640 | 1,726.5 (21.58%) | 1,349.1 (16.86%) | 150.1 (1.88%) |
| 2,048 | 12,288 | 251,658,240 | 2,822.8 (35.29%) | 2,822.3 (35.28%) | 332.4 (4.15%) |
| 8,192 | 12,288 | 1,006,632,960 | 3,920.4 (49.01%) | 3,922.4 (49.03%) | 326.8 (4.09%) |

The eager path uses the same logical-byte denominator for comparison, but it
performs additional physical traffic for FP32 casts, intermediates, separate
pointwise kernels, concatenation, and the final BF16 cast. Its reported value
is useful/logical bandwidth, not measured HBM traffic. The local GPU is exposed
as a virtual function, and the 8.0 TB/s figure is an advertised theoretical
full-GPU peak rather than sustained application bandwidth.

### Roofline estimate

Per gate/up pair, the fused kernel transfers 10 logical bytes and performs
approximately 11 ordinary FP32 operations plus one `exp2`, or about
**1.1 FLOP/byte** when the transcendental is excluded from the FLOP count.
AMD publishes 144.2 TFLOP/s peak FP32 vector performance and 8.0 TB/s peak HBM
bandwidth for MI350X, giving an estimated roofline ridge point of
`144.2 / 8.0 = 18.0 FLOP/byte`.

Because 1.1 FLOP/byte is far below 18.0 FLOP/byte, the conventional roofline
predicts a bandwidth-bound kernel at sufficiently large shapes. The largest
cases reach about 39-49% of advertised peak HBM bandwidth. This is an estimate:
`exp2` throughput, non-payload traffic, virtualization, and QoS are not modeled.

## Validation

MI350X/gfx950 targeted correctness, including width 1, strided gradients,
explicit output, empty rows, unsupported gfx942 config behavior, and a
non-current-GPU launch:

```bash
GPU_ARCHS=gfx950 PYTHONPATH=. pytest -q \
  op_tests/triton_tests/fusions/test_silu_and_mul_backward.py
```

Result: `53 passed`.

Kernel and config-loader regression suite:

```bash
GPU_ARCHS=gfx950 PYTHONPATH=. pytest -q \
  op_tests/triton_tests/fusions/test_silu_and_mul_backward.py \
  op_tests/triton_tests/fusions/test_fused_clamp_act_mul.py \
  op_tests/triton_tests/normalization/test_fused_add_rmsnorm_pad.py
```

Result: `1223 passed, 874 skipped`.

Activation-family regression suite:

```bash
GPU_ARCHS=gfx950 PYTHONPATH=. pytest -q \
  op_tests/triton_tests/test_activation.py \
  op_tests/triton_tests/fusions/test_fused_silu_mul.py
```

Result: `419 passed, 228 skipped`.

Width-one benchmark smoke test:

```bash
PYTHONPATH=. python \
  op_tests/op_benchmarks/triton/bench_silu_and_mul_backward.py \
  --rows 3 --width 1 --provider all --metric time --warmup 5 --rep 20
```

Result: all providers completed.

VGPR path:

```bash
PYTHONPATH=. python \
  op_tests/op_benchmarks/triton/bench_silu_and_mul_backward.py \
  --rows 1 --width 64 --provider aiter --metric time \
  --warmup 1 --rep 1 --print-vgpr
```

Result: `_silu_and_mul_backward_kernel_BLOCK_M_1_BLOCK_N_64`, 11 VGPR, no
VGPR spill.

Ruff 0.16.0, Black 26.5.1, `compileall`, JSON parsing, and
`git diff --check` pass for all changed files.

## Documentation, dependencies, and compatibility

- The public wrapper documents the config parameters and gfx950-only support.
- The shared config selector is documented in `configs/CLAUDE.md`, the Triton
  maintainer guide, and the scoped review instructions.
- No new third-party dependency.
- No breaking change to an existing public API.
- The forward path remains the existing `aiter.silu_and_mul` operation.

### CODE REVIEW

File: aiter/ops/triton/fusions/silu_and_mul_backward.py [hunk 1]
```diff
@@ -0,0 +1,106 @@
...
+
+def _get_config(n_cols: int) -> dict:
+    config_dir = resolve_config_dir("fusions", "SILU-AND-MUL-BACKWARD")
+    configs = load_config_json(f"{config_dir}/DEFAULT.json", required=True)
+    for bound in sorted(
+        int(key.removeprefix("N_LEQ_")) for key in configs if key.startswith("N_LEQ_")
+    ):
+        if n_cols <= bound:
+            return dict(configs[f"N_LEQ_{bound}"])
+    return dict(configs["any"])
```
[the remark this answers is not in the document]
Comment (DaiXindi-AMD) [reply aiter/ops/triton/fusions/silu_and_mul_backward.py]: Fixed in 9692c5729: added shared select_leq_config in config_utils and migrated silu_and_mul_backward, fused_clamp_act_mul, and fused_add_rmsnorm_pad to it. The public docstring now documents N_LEQ threshold dispatch and BLOCK_M/BLOCK_N/num_warps/num_stages. Tests: backward 45 passed; RMSNorm 840 passed/420 skipped; fused clamp 330 passed/454 skipped.

**Comment by DaiXindi-AMD** [comment]
Follow-up: the same review fix is now at HEAD 214d1c6f5. The amend only adds the README and review-instruction updates required when introducing a shared config-selection convention; the tested kernel/helper behavior is unchanged.
